libs/libcommon/include/common/DateLUTImpl.h

```diff
@@ -352,54 +313,205 @@ class DateLUTImpl
+        releveant if WEEK_JANUARY is not set.
+
+                  If set	 Week is in range 1-53.
+
+        In this case one may get week 53 for a date in January (when
+        the week is that last week of previous year) and week 1 for a
+        date in December.
+
+      WEEK_FIRST_WEEKDAY (2)  If not set	Weeks are numbered according
+                        to ISO 8601:1988
+                  If set	The week that contains the first
+                        'first-day-of-week' is week 1.
+
+      WEEK_NEWYEAR_DAY (3)
+                  If set	The week that contains the January 1 is week 1.
+                            Week is in range 1-53.
+                            And ignore WEEK_YEAR, WEEK_FIRST_WEEKDAY
+
+        ISO 8601:1988 means that if the week containing January 1 has
+        four or more days in the new year, then it is week 1;
+        Otherwise it is the last week of the previous year, and the
+        next week is week 1.
+    */
+    inline unsigned calc_week(DayNum d, UInt32 week_mode, UInt32 * year) const
+    {
+        bool newyear_day_mode = week_mode & WEEK_NEWYEAR_DAY;
+
+        week_mode = check_week_mode(week_mode);
+        bool monday_first_mode = week_mode & WEEK_MONDAY_FIRST;
+        bool week_year_mode = week_mode & WEEK_YEAR;
+        bool first_weekday_mode = week_mode & WEEK_FIRST_WEEKDAY;
+
+        // Calculate week number of WEEK_NEWYEAR_DAY mode
+        if (newyear_day_mode)
+        {
+            return calc_newyear_week(d, monday_first_mode, year);
+        }
+
+        UInt32 days = 0;
+        UInt64 daynr = calc_daynr(toYear(d), toMonth(d), toDayOfMonth(d));
+        UInt64 first_daynr = calc_daynr(toYear(d), 1, 1);
+
+        // 0 for monday, 1 for tuesday ...
+        // get weekday from first day in year.
+        UInt32 weekday = calc_weekday(first_daynr, !monday_first_mode);
+        *year = toYear(d);
+
+        if (toMonth(d) == 1 && toDayOfMonth(d) <= 7 - weekday)
+        {
+            if (!week_year_mode && ((first_weekday_mode && weekday != 0) || (!first_weekday_mode && weekday >= 4)))
+                return 0;
+            week_year_mode = 1;
+            (*year)--;
+            first_daynr -= (days = calc_days_in_year(*year));
+            weekday = (weekday + 53 * 7 - days) % 7;
+        }
+
+        if ((first_weekday_mode && weekday != 0) || (!first_weekday_mode && weekday >= 4))
+            days = daynr - (first_daynr + (7 - weekday));
+        else
+            days = daynr - (first_daynr - weekday);
+
+        if (week_year_mode && days >= 52 * 7)
+        {
+            weekday = (weekday + calc_days_in_year(*year)) % 7;
+            if ((!first_weekday_mode && weekday < 4) || (first_weekday_mode && weekday == 0))
+            {
+                (*year)++;
+                return 1;
+            }
+        }
+        return days / 7 + 1;
     }
 
-    inline unsigned toISOWeek(time_t t) const
+    inline unsigned calc_yearWeek(DayNum d, UInt32 week_mode) const
```

Comment:
It looks like you don't need two functions `calc_week()` and `calc_yearWeek()` because you have to calculate `year` anyway. So it seems it's enough to have only function

```
std::pair<UInt16 /* year */, UInt8 /* week */> toYearAndWeek(DayNum d, WeekMode week_mode) const
```

and the `WeekImpl::execute()` and `YearWeekImpl::execute()` functions get from this pair everything they need.

Reply:
ok, I'll fix it.